libraries/AP_RCProtocol/AP_RCProtocol_SRXL2.cpp

```diff
@@ -0,0 +1,310 @@
+        AP_HAL::panic("Duplicate SRXL2 handler");
+    }
+
+    _singleton = this;
+    // Init the local SRXL device
+    if (!srxlInitDevice(SRXL_DEVICE_ID, SRXL_DEVICE_PRIORITY, SRXL_DEVICE_INFO, uniqueID)) {
+        AP_HAL::panic("Failed to initialize SRXL2 device");
+    }
+
+    // Init the SRXL bus: The bus index must always be < SRXL_NUM_OF_BUSES -- in this case, it can only be 0
+    if (!srxlInitBus(0, this, SRXL_SUPPORTED_BAUD_RATES)) {
+        AP_HAL::panic("Failed to initialize SRXL2 bus");
+    }
+
+}
+
+void AP_RCProtocol_SRXL2::process_pulse(uint32_t width_s0, uint32_t width_s1)
+{
+    uint8_t b;
+    if (ss.process_pulse(width_s0, width_s1, b)) {
+        _process_byte(ss.get_byte_timestamp_us(), b);
+    }
+}
+
+void AP_RCProtocol_SRXL2::_process_byte(uint32_t timestamp_us, uint8_t byte)
+{
+    if (_decode_state == STATE_IDLE) {
+        switch (byte) {
+        case SPEKTRUM_SRXL_ID:
+            _frame_len_full = 0U;
+            _decode_state = STATE_NEW;
+            break;
+        default:
+            _frame_len_full = 0U;
+            _decode_state = STATE_IDLE;
+            _buflen = 0;
+            return;
+        }
+    }
+
+    switch (_decode_state) {
+    case STATE_NEW:  // buffer header byte and prepare for frame reception and decoding
+        _buffer[0U]=byte;
+        _buflen = 1U;
+        _decode_state_next = STATE_COLLECT;
+        break;
+
+    case STATE_COLLECT: // receive all bytes. After reception decode frame and provide rc channel information to FMU
+        _buffer[_buflen] = byte;
+        _buflen++;
+
+        // need a header to get the length
+        if (_buflen < SRXL2_HEADER_LEN) {
+            return;
+        }
+
+        // parse the length
+        if (_buflen == SRXL2_HEADER_LEN) {
+            _frame_len_full = _buffer[2];
+            return;
+        }
+
+        if (_buflen > _frame_len_full) {
+            // a logic bug in the state machine, this shouldn't happen
+            _decode_state = STATE_IDLE;
+            _buflen = 0;
+            _frame_len_full = 0;
+            return;
+        }
+
+        if (_buflen == _frame_len_full) {
+            log_data(AP_RCProtocol::SRXL2, timestamp_us, _buffer, _buflen);
+            // Try to parse SRXL packet -- this internally calls srxlRun() after packet is parsed and resets timeout
+            if (srxlParsePacket(0, _buffer, _frame_len_full)) {
+                add_input(_max_channels, _channels, _in_failsafe, _new_rssi);
+            }
+            _last_run_ms = AP_HAL::millis();
+            _decode_state_next = STATE_IDLE;
+        } else {
+            _decode_state_next = STATE_COLLECT;
+        }
+        break;
+
+    default:
+        break;
+    }
+
+    _decode_state = _decode_state_next;
+    _last_data_us = timestamp_us;
+}
+
+void AP_RCProtocol_SRXL2::update(void)
+{
+    // it's not clear this is actually required, perhaps on power loss?
+    if (frontend.protocol_detected() == AP_RCProtocol::SRXL2) {
```

Comment:
why are we checking what protocol the frontend is using?

Reply:
It was my attempt to see if this was causing the FC lockups I am seeing, but it is not. Disabling this code makes no difference. Not writing to the UART definitely does however so it's either to do with the state machine or with our half-duplex support.

Reply:
Actually this was intentional - we call srxlRun() in case the state machine needs updating when we have no data, but it makes no sense to do this if we don't think we are decoding  SRXL2.

Reply:
Ok fixed